repose-aggregator/commons/utilities/src/main/scala/org/openrepose/commons/utils/servlet/http/HttpServletRequestWrapper.scala

```diff
@@ -0,0 +1,150 @@
+/*
+ * _=_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_=
+ * Repose
+ * _-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-
+ * Copyright (C) 2010 - 2015 Rackspace US, Inc.
+ * _-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * =_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_=_
+ */
+package org.openrepose.commons.utils.servlet.http
+
+import java.io.{InputStreamReader, BufferedReader}
+import java.util
+import javax.servlet.ServletInputStream
+import javax.servlet.http.HttpServletRequest
+
+import org.apache.http.client.utils.DateUtils
+
+import scala.collection.JavaConverters._
+import scala.collection.immutable.{TreeMap, TreeSet}
+
+/**
+ * Created with IntelliJ IDEA.
+ * User: adrian
+ * Date: 5/27/15
+ * Time: 10:25 AM
+ */
+class HttpServletRequestWrapper(originalRequest: HttpServletRequest, inputStream: ServletInputStream)
+  extends javax.servlet.http.HttpServletRequestWrapper(originalRequest)
+  with HeaderInteractor {
+
+  object RequestBodyStatus extends Enumeration {
+    val Available, InputStream, Reader = Value
+  }
+
+  private var status = RequestBodyStatus.Available
+
+  def this(originalRequest: HttpServletRequest) = this(originalRequest, originalRequest.getInputStream())
+
+  val caseInsensitiveOrdering = Ordering.by[String, String](_.toLowerCase)
+
+  private var headerMap: Map[String, List[String]] = new TreeMap[String, List[String]]()(caseInsensitiveOrdering)
+  private var removedHeaders: Set[String] = new TreeSet[String]()(caseInsensitiveOrdering)
+
+  override def getInputStream: ServletInputStream = {
+    if (status == RequestBodyStatus.Reader) throw new IllegalStateException else status = RequestBodyStatus.InputStream
+    inputStream
+  }
+
+  override def getReader: BufferedReader = {
+    if (status == RequestBodyStatus.InputStream) throw new IllegalStateException else status = RequestBodyStatus.Reader
+    new BufferedReader(new InputStreamReader(inputStream))
+  }
+
+  def getHeaderNamesScala: Set[String] = headerMap.keySet ++ super.getHeaderNames.asScala.toSet.filterNot(removedHeaders.contains)
```

Comment:
I'm not overly concerned with the security contract. If we get prevented by the container on our super call that's fine. If they've added one under that name they probably want to be able to access it, so it should work as is. The null thing is freaking stupid and i hate the damn contract. We'll end up wrapping the reques tin an option and accessing it out that way.